gpus/python-gpu-example.html.md

```diff
@@ -0,0 +1,242 @@
+app = "cgpu-allinone"   # Change this to your app's name
+primary_region = "ord"  # If you change this, ensure it's to a region that offers GPUs
+vm.size = "a100-40gb"   # A shorthand for the size preset in the [[vm]] section
+swap_size_mb = 32768    # This enables 32GB swap
+
+[build]
+  [build.args]
+    NONROOT_USER = "pythonuser" # Access this value in the Dockerfile using `ARG NONROOT_USER`
+
+# Use a volume to store LLMs or any big file that doesn't fit in a Docker image
+# This whole volume will be the non-root user's home directory
+[mounts]
+source = "data"
+destination = "/home/pythonuser"   # Make sure this matches the value of the NONROOT_USER build arg
+# initial_size = "50gb"            # Uncomment to set the size for the volume created on first deployment
+```
+
+### The Docker image
+
+Ready-made Docker images exist for many ML-related projects. There is a CUDA-enabled Jupyter Docker image, but it's kind of a black box and it's not maintained by the Jupyter folks, though they link to it. 
+
+Here we'll go step by step from a small Ubuntu image, installing and configuring a Python development environment that can use the GPU's CUDA capabilities, kind of like we would with a normal computer.
+
+Here's a summary of what the Dockerfile does;
+
+* Uses Ubuntu 22.04 as a base image, as it's compatible with the NVIDIA drivers on Fly Machines.
+* Installs system-wide packages with the apt package manager: python3, python3-pip, python3-venv, python3-wheel, git, nano (substitute your favourite terminal-compatible text editor here).
+* Adds a non-root user to own the Python venv and run things. Gets the name for this user from a `[build.arg]` set in `fly.toml`.
+* Copies the scripts for root and the user to run at startup.
+* CMD invokes the first script with the non-root user name as argument.
+
+```Dockerfile
+FROM ubuntu:22.04
+RUN apt update -q && apt install -y python3 python3-pip python3-venv python3-wheel git nano && \
+    apt clean && rm -f /var/lib/apt/lists/*_*
+
+ARG NONROOT_USER
+RUN echo "User will be $NONROOT_USER"
+ENV PYTHON_USER=$NONROOT_USER
+
+# Create unprivileged user with a home dir and using bash
+RUN useradd -ms /bin/bash $PYTHON_USER
+
+COPY --chmod=0755 ./entrypoint.sh ./entrypoint.sh
+COPY --chown=$PYTHON_USER:$PYTHON_USER --chmod=0755 ./post-initialization.sh ./post-initialization.sh
+# If you have a requirements.txt for the project, uncomment this and
+# adjust post-initialization.sh to use it
+# COPY --chown=$PYTHON_USER:$PYTHON_USER requirements.txt .
+
+# CMD ["sleep", "inf"]
+CMD ["/bin/bash", "-c", "./entrypoint.sh $PYTHON_USER"]
+```
+
+### Entrypoint script
+When this Machine boots, it runs the script `entrypoint.sh` as root. 
+
+This script gives the non-root user (whose username is set to `pythonuser` via a build argument in `fly.toml`) ownership of the non-root user's home directory to that user. This is necessary, because we're mounting a Fly Volume over that point in the Machine file system.
```

Comment:
```suggestion
This script gives the non-root user (whose username is set to `pythonuser` via a build argument in `fly.toml`) ownership of the non-root user's home directory. This is necessary, because we're mounting a Fly Volume over that point in the Machine file system.
```